EventFilter/RawDataCollector/src/RawDataMapperByLabel.cc

```diff
@@ -0,0 +1,113 @@
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+#include <iostream>
+
+using namespace edm;
+
+class RawDataMapperByLabel: public edm::stream::EDProducer<> {
+public:
+    
+    ///Constructor
+    RawDataMapperByLabel(const edm::ParameterSet& pset);
+    
+    ///Destructor
+    ~RawDataMapperByLabel() override;
+    
+    void produce(edm::Event & e, const edm::EventSetup& c) override; 
+    
+    static void fillDescriptions(edm::ConfigurationDescriptions &);     
+private:
+    
+    typedef std::vector<edm::InputTag>::const_iterator tag_iterator_t;
+    typedef std::vector<edm::EDGetTokenT<FEDRawDataCollection> >::const_iterator tok_iterator_t;
+
+    std::vector<edm::InputTag> inputTags_ ;
+    std::vector<edm::EDGetTokenT<FEDRawDataCollection> > inputTokens_;
+    
+    bool firstEvent_;
+    edm::InputTag filledCollectionName_;
+    edm::InputTag mainCollectionTag_ ;
+    
+    
+
+};
+
+
+RawDataMapperByLabel::RawDataMapperByLabel(const edm::ParameterSet& pset) {
+
+  inputTags_ = pset.getParameter<std::vector<InputTag> >("rawCollectionList");
+  mainCollectionTag_ = pset.getParameter<InputTag>("rainCollection");
+  firstEvent_= true;
+  filledCollectionName_= InputTag("");
+
+  inputTokens_.reserve(inputTags_.size());
+  for(auto const& inputTag: inputTags_) {
+    inputTokens_.push_back(consumes<FEDRawDataCollection>(inputTag));
+  }
+  produces<FEDRawDataCollection>();
+ 
+}
+
+RawDataMapperByLabel::~RawDataMapperByLabel(){
+
+}
+
+
+void RawDataMapperByLabel::produce(Event & e, const EventSetup& c){
+ 
+ bool alreadyACollectionFilled= false;
+ tag_iterator_t inputTag = inputTags_.begin();
+ for(tok_iterator_t inputTok = inputTokens_.begin(); inputTok != inputTokens_.end(); ++inputTok, ++inputTag  ) {
+   Handle<FEDRawDataCollection> input;
+   if (e.getByToken(*inputTok,input)){
+    	if(input.isValid()){
+    	    if(firstEvent_){  
+    	    	filledCollectionName_ = *inputTag; 
+    	    	alreadyACollectionFilled = true;
+                firstEvent_= false;  
+            }
+            
+    		if(alreadyACollectionFilled) throw cms::Exception("Unknown input type") << "Two input collections are present. Please make sure that the input dataset has only one FEDRawDataCollector collection filled";
+            if(!(filledCollectionName_==*inputTag)) throw cms::Exception("Unknown input type") << "The filled collection has changed!";
+            
+            if(!(mainCollectionTag_==filledCollectionName_)) e.put(std::make_unique<FEDRawDataCollection>(*input.product()));
```

Comment:
this is to revive https://github.com/cms-sw/cmssw/pull/24819#discussion_r223841035
the same problem persists here.
I'm afraid, there has to be a ```put``` in every event.

@Dr15Jones can we expect a valid behavior downstream if moduleA in the current process checks for an input moduleA from the source and doesn't event::put anything even though it declares that it produces; so that in this case the consumers will correctly pick up data labeled moduleA from the source?

Reply:
ah, good.
For some reason I thought that an EDGetToken is rigidly pointing to a known existing producer.